enhancements/update/change-management-and-maintenance-schedules.md

```diff
@@ -0,0 +1,1186 @@
+is always permissive -- allowing material changes to be initiated (see [Change Management
+Hierarchy](#change-management-hierarchy) for caveats).
+
+All change management strategies, except `Disabled`, are subject to the following `changeManagement` fields:
+- `changeManagement.disabledUntil: <bool|date>`: When `disabledUntil: true` or `disabledUntil: <future-date>`, the interpreted strategy for 
+  change management in the resource is `Disabled`. Setting a future date in `disabledUntil` offers a less invasive (i.e. no important configuration needs to be changed) method to 
+  disable change management constraints (e.g. if it is critical to roll out a fix) and a method that
+  does not need to be reverted (i.e. it will naturally expire after the specified date and the configured
+  change management strategy will re-activate).
+- `changeManagement.pausedUntil: <bool|date>`: Unless the effective active strategy is Disabled, `pausedUntil: true` or `pausedUntil: <future-date>`, change management must 
+  pause material changes.
+
+### Change Management Status
+Change Management information will also be reflected in resource status. Each resource 
+which contains the stanza in its `spec` will expose its current impact in its `status`. 
+Common user interfaces for aggregating and displaying progress of these underlying resources
+should be updated to proxy that status information to the end users.
+
+### Change Management Metrics
+Cluster wide change management information will be made available through cluster metrics. Each resource
+containing the stanza must expose the following metrics:
+- The number of seconds until the next known permitted change window. 0 if changes can currently be initiated. -1 if changes are paused indefinitely. -2 if no permitted window can be computed.
+- Whether any change management strategy is enabled.
+- Which change management strategy is enabled.
+- If changes are pending due to change management controls.
+
+### Change Management Hierarchy
+Material changes to worker-nodes are constrained by change management policies in their associated resource AND 
+at the control-plane resource. For example, in a standalone profile, if a MachineConfigPool's change management
+configuration apparently permits material changes from being initiated at a given moment, that is only the case 
+if ClusterVersion is **also** permitting changes from being initiated at that time.
+
+The design choice is informed by a thought experiment: As a cluster lifecycle administrator for a Standalone cluster,
+who wants to achieve the simple goal of ensuring no material changes take place outside a well-defined 
+maintenance schedule, do you want to have to the challenge of keeping every MachineConfigPool's 
+`changeManagement` stanza in perfect synchronization with the ClusterVersion's? What if a new MCP is created 
+without your knowledge?

```

Comment:
MachineConfigPools consume inputs from many places, and ClusterVersion updates are only one aspect.  https://amd64.ocp.releases.ci.openshift.org/ -> [4.16.0-ec.4](https://amd64.ocp.releases.ci.openshift.org/releasestream/4-dev-preview/release/4.16.0-ec.4) -> [aws-ovn-serial](https://prow.ci.openshift.org/view/gs/test-platform-results/logs/periodic-ci-openshift-release-master-nightly-4.16-e2e-aws-ovn-serial/1767648317749071872) -> [Artifacts](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/logs/periodic-ci-openshift-release-master-nightly-4.16-e2e-aws-ovn-serial/1767648317749071872/) -> ... -> [gather-extra artifacts](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/logs/periodic-ci-openshift-release-master-nightly-4.16-e2e-aws-ovn-serial/1767648317749071872/artifacts/e2e-aws-ovn-serial/gather-extra/artifacts/):

```console
$ curl -s https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/logs/periodic-ci-openshift-release-master-nightly-4.16-e2e-aws-ovn-serial/1767648317749071872/artifacts/e2e-aws-ovn-serial/gather-extra/artifacts/machineconfigs.json | jq -r '.items[].metadata | .creationTimestamp + " " + .name' | grep rendered-worker- | sort
2024-03-12T20:45:43Z rendered-worker-9724b3c3c79d2e58b093eb46d0c2c3ff
2024-03-12T20:51:58Z rendered-worker-33a6a11e659aca048b955f168eb770f3
2024-03-12T20:52:04Z rendered-worker-3e2eab9d46a1db0948347f34d268ea61
$ curl -s https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/logs/periodic-ci-openshift-release-master-nightly-4.16-e2e-aws-ovn-serial/1767648317749071872/artifacts/e2e-aws-ovn-serial/gather-extra/artifacts/machineconfigpools.json | jq -c '.items[] | select(.metadata.name == "worker").status.configuration.source[]'
{"apiVersion":"machineconfiguration.openshift.io/v1","kind":"MachineConfig","name":"00-worker"}
{"apiVersion":"machineconfiguration.openshift.io/v1","kind":"MachineConfig","name":"01-worker-container-runtime"}
{"apiVersion":"machineconfiguration.openshift.io/v1","kind":"MachineConfig","name":"01-worker-kubelet"}
{"apiVersion":"machineconfiguration.openshift.io/v1","kind":"MachineConfig","name":"97-worker-generated-kubelet"}
{"apiVersion":"machineconfiguration.openshift.io/v1","kind":"MachineConfig","name":"98-worker-generated-kubelet"}
{"apiVersion":"machineconfiguration.openshift.io/v1","kind":"MachineConfig","name":"99-worker-generated-registries"}
{"apiVersion":"machineconfiguration.openshift.io/v1","kind":"MachineConfig","name":"99-worker-ssh"}
```

Instead of having it implicitly assumed that ClusterVersion is going to be the controlling resource taking precedence over local MachineConfigPool configuration, maybe `changeManagement` could explicitly declare that with references to higher-precedence managers?

Or, because that would make the system even more complicated, how likely is it that folks are creating new MachineConfigPools without looping in cluster lifecycle administrators?  If cluster lifecycle administrators cannot agree amongst themselves, that is not a problem we can solve with tooling.  If they do agree, but want unified control over the policy, I'm not as down on UpdatePolicy CRD as the current alternatives-section is, and possibly there's a way to mitigate the concerns around that while still avoiding implicit or explicit hierarchy complications.

Reply:
There were two motivations to the current proposal:
1. It is intended to be implementable with existing CAPI primitives. CAPI exposes "paused" on the two main resources [Cluster](https://pkg.go.dev/sigs.k8s.io/cluster-api/api/v1alpha3#ClusterSpec) and [MachineDeployment](https://pkg.go.dev/sigs.k8s.io/cluster-api/api/v1alpha3#MachineDeploymentPhase). My understanding is that pausing the Cluster pauses the MachineDeployment but not vice versa. This should allow the HyperShift operator to fulfill the simple changeManagement hierarchy by  pausing and unpausing these resources. If the enhancement calls for independence of changeManagement stanzas, we would need to be confident CAPI would accept a significant paradigm shift. I would only feel compelled to do that if we thought there was a significant advantage.
2. The guiding principle of "simple things should be simple." Setting something globally, once, and not having to keep it in mind for an unrelated activity reduces the mental overhead associated with the product. A diligent enough team can always solve for <footgun> with policy / clever automation / etc. But the more we ask this theoretical team to solve, the less pleasant the product is to use. "Cluster lifecycle administrator" is often going to be a single person trying to do lifecycle management along with everything else. Ideally, the enhancement will improve their QOL while also being flexible enough for plausible complicated use cases.